### ISSUE TITLE
"pydra.engine.specs.Path" type not satisfiable when type checking is switched on

### ISSUE THREAD

**Issue by tclose** [nipype/pydra#579]
What version of Pydra are you using?

pydra==0.19
pydra-mrtrix3==0.1.1

What were you trying to do?

Use an interface with a `Path` input (e.g. MRConvert from pydra-mrtrix3) within a workflow when "set_input_validator" is set to True

What did you expect will happen?

I would be able to pass a Path (or str)-typed variable to it

What actually happened?

The Path variables get automatically converted to str by `ShellSpec.retrieve_values()` and then fails type validation for Path type

Can you replicate the behavior? If yes, how?

```python
from pathlib import Path
from pydra import set_input_validator
set_input_validator(True)
from pydra import Workflow
from pydra.tasks.mrtrix3.utils import MRConvert


wf = Workflow(name="path_mwe", input_spec=["in_file"])

wf.add(MRConvert(
    name="mrconvert",
    in_file=wf.lzin.in_file,
    out_file='out_file.mif'))

wf.set_output([('out_file', wf.mrconvert.lzout.out_file)])

wf(in_file=Path('/path/to/input/nifti.nii.gz'))

wf()

```

**Comment by satra** [comment]
indeed, any such operation should not change the state of attributes of the object. the `retrieve_values` is being used in the workflow run function to gather the lazy outputs. 

what happens if you comment out this section that converts `Path` to `str`: https://github.com/nipype/pydra/blob/master/pydra/engine/specs.py#L377

i'm assuming it will break somewhere and some tests and those may be the places to fix.

**Comment by tclose** [comment]
The tests actually pass when that line is commented out (I'm assuming some code will fail somewhere though), presumably because the inputs are passed as strings to begin with.

However, when you `set_input_validators(True)` then you get type errors in some of the shell command tests, e.g.

```
TypeError: ("'orig_file' must be (<class 'str'>, <class 'bytes'>, <class 'attr._make._Nothing'>) (got PosixPath('/private/var/folders/mz/yn83q2fd3s758w1j75d2nnw80000gn/T/pytest-of-tclose/pytest-10/test_wf_shell_cmd_ndst_1_cf_0/
```

I presumed this was because they define the file inputs as `str` rather than `Path`, so I converted the input spect of the  test task to use `Path` instead, but then I get the dreaded `Exception: graph is not empty, but not able to get more tasks - something is wrong (e.g. with the filesystem)` error, which occurs quite often when there is something structurally wrong with the workflow.

**Comment by tclose** [comment]
Sorry, I meant in the test_wf_shell_cmd3a, test_wf_shel_cmd_ndist_1 and test_wf_shell_cmd_state_1 unittests

**Comment by djarecka** [comment]
I will try to recall why I did the transformation and perhaps add more test that use the validator... Thank you for checking this!
